src/fiber/fiber.ml

```diff
@@ -453,6 +453,56 @@ module Mutex = struct
   let create () = { locked = false; waiters = Queue.create () }
 end
 
+module Throttle = struct
+  type t =
+    { mutable size : int
+    ; mutable running : int
+    ; waiting : unit Ivar.t Queue.t
+    }
+
+  let create size = { size; running = 0; waiting = Queue.create () }
+
+  let size t = t.size
+
+  let running t = t.running
+
+  let restart t =
+    if Queue.is_empty t.waiting || t.running >= t.size then
+      return ()
+    else
+      let ivar = Queue.pop_exn t.waiting in
+      t.running <- t.running + 1;
+      Ivar.fill ivar ()
+
+  let resize t n =
+    t.size <- n;
+    restart t
```

Comment:
This will resume at most one blocked job. One way to fix this to make restart should call itself recursively in the `else` clause`

Reply:
Indeed. I also removed the thing about fairness in `run`, that shouldn't be necessary once we change the order in which ivar does things.